src/openstack_cloud/openstack_manager.py

```diff
@@ -1172,3 +1249,222 @@ def flush(self) -> int:
                 remove_token=remove_token,
             )
             return len(runners_to_delete)
+
+    def _pull_metrics(self, server: Server, instance_name: str) -> None:
+        """Pull metrics from the runner into the respective storage for the runner.
+
+        Args:
+            server: The Openstack server instance.
+            instance_name: The Openstack server name.
+        """
+        try:
+            storage = metrics_storage.get(instance_name)
+        except GetMetricsStorageError:
+            logger.exception(
+                "Failed to get shared metrics storage for runner %s, "
+                "will not be able to issue all metrics.",
+                instance_name,
+            )
+            return
+
+        for ssh_conn in self._get_ssh_connections(server=server):
+            try:
+                self._pull_file(
+                    ssh_conn=ssh_conn,
+                    remote_path=str(METRICS_EXCHANGE_PATH / "pre-job-metrics.json"),
+                    local_path=str(storage.path / "pre-job-metrics.json"),
+                    max_size=MAX_METRICS_FILE_SIZE,
+                )
+                self._pull_file(
+                    ssh_conn=ssh_conn,
+                    remote_path=str(METRICS_EXCHANGE_PATH / "post-job-metrics.json"),
+                    local_path=str(storage.path / "post-job-metrics.json"),
+                    max_size=MAX_METRICS_FILE_SIZE,
+                )
+                return
+            except _PullFileError as exc:
+                logger.warning(
+                    "Failed to pull metrics for %s: %s . Will not be able to issue all metrics",
+                    instance_name,
+                    exc,
+                )
+                return
+            except _SSHError as exc:
+                logger.info("Failed to pull metrics for %s: %s", instance_name, exc)
+                continue
```

Comment:
The downside here is that if only the `pre-job-metrics.json` is pulled, the runner is considered crashed (see logic in `_issue_reconciliation_metric`), we may want to revisit this. But for now I kept it the same as for LXD.